src/Microsoft.Security.Utilities.Core/CachedDotNetRegex.cs

```diff
@@ -35,6 +37,16 @@ public static Regex GetOrCreateRegex(string pattern, RegexOptions options)
 #endif
         }
 
+        internal static string NormalizeGroupsPattern(string pattern)
+        {
+            if (pattern.IndexOf("?P<") != -1)
```

Comment:
I'm sad to see this return because it showed up in profiling. 

In general, it's never great when allocations are done to look something up in a dictionary. 

At a bare minimum, we don't need two passes over the pattern for this normalization. string.Replace returns `this` when no match is found. The `if` is doubling the cost of the replacement.

Reply:
yes good catch will fix that and then let's take your suggestion.